File: GraffiTab-iOS/Classes/ViewControllers/Main/Main/UserProfile/Edit/EditProfileViewController.m
//
//  EditProfileViewController.m
//  GraffiTab-iOS
//
//  Created by Georgi Christov on 17/04/2015.
//  Copyright (c) 2015 GraffiTab. All rights reserved.
//

#import "EditProfileViewController.h"
#import "FXBlurView.h"
#import "UIActionSheet+Blocks.h"
#import "EXPhotoViewer.h"
#import "UIWindow+PazLabs.h"
#import "EditAvatarTask.h"
#import "EditCoverTask.h"
#import "ImageCropViewController.h"
#import "EditTextFieldViewController.h"
#import "EditTextViewViewController.h"
#import "EditProfileTask.h"

#define IMAGE_PICKER_MODE_AVATAR 0
#define IMAGE_PICKER_MODE_COVER 1

@interface EditProfileViewController () {
    
    IBOutlet UIImageView *avatarImage;
    IBOutlet UIImageView *coverImage;
    
    IBOutlet UILabel *firstNameLabel;
    IBOutlet UILabel *lastNameLabel;
    IBOutlet UILabel *emailLabel;
    IBOutlet UILabel *aboutLabel;
    IBOutlet UILabel *websiteLabel;
    
    int imagePickerMode;
    Person *user;
    UIImage *defaultBlurredImage;
    UIImagePickerController *galleryPicker;
}

@end

@implementation EditProfileViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    user = [Settings getInstance].user;
    
    [self setupTopBar];
    [self setupDefaultImage];
    [self setupImageViews];
    
    [self loadData];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    if (self.navigationController.isNavigationBarHidden)
        [self.navigationController setNavigationBarHidden:NO animated:YES];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)onClickSave:(id)sender {
    NSString *fn = firstNameLabel.text;
    NSString *ln = lastNameLabel.text;
    NSString *em = emailLabel.text;
    NSString *ab = aboutLabel.text;
    NSString *we = websiteLabel.text;
    
    if ([InputValidator validateProfileEditInput:fn lastName:ln email:em about:ab website:we viewController:self.navigationController]) {
        [[LoadingViewManager getInstance] addLoadingToView:self.navigationController.view withMessage:@"Processing"];
        
        EditProfileTask *task = [EditProfileTask new];
        [task editProfileWithFirstName:fn lastName:ln email:em about:ab website:we successBlock:^(ResponseObject *response) {
            [[LoadingViewManager getInstance] removeLoadingView];
            
            user = [Settings getInstance].user;
            
            [self loadData];
            
            SCLAlertView *alert = [[SCLAlertView alloc] init];
            [alert showTitle:self.navigationController title:APP_NAME subTitle:@"Profile saved!" style:Success closeButtonTitle:@"OK" duration:0.0f];
        } failureBlock:^(ResponseObject *response) {
            [[LoadingViewManager getInstance] removeLoadingView];
            
            if (response.reason == ALREADY_EXISTS)
                [[SCLAlertView new] showError:self.navigationController title:APP_NAME subTitle:@"This username or email have already been taken." closeButtonTitle:@"OK" duration:0.0f];
            else if (response.reason == AUTHORIZATION_NEEDED) {
                [Utils logoutUserAndShowLoginController];
                [Utils showMessage:APP_NAME message:@"Your session has timed out. Please login again."];
            }
            else if (response.reason == DATABASE_ERROR || response.reason == NOT_FOUND || response.reason == NETWORK || response.reason == OTHER)
                [Utils showMessage:APP_NAME message:response.message];
        }];
    }
}

- (void)onClickImage {
    NSArray *actions = @[@"View image", @"Change image", @"Remove image"];
    
    [UIActionSheet showInView:self.view
                    withTitle:@"What would you like to do?"
            cancelButtonTitle:@"Cancel"
       destructiveButtonTitle:nil
            otherButtonTitles:actions
                     tapBlock:^(UIActionSheet *actionSheet, NSInteger buttonIndex) {
                         if ([[actionSheet buttonTitleAtIndex:buttonIndex] isEqualToString:@"Cancel"])
                             return;
                         
                         if ([[actionSheet buttonTitleAtIndex:buttonIndex] isEqualToString:@"View image"]) {
                             dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0.5 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
                                 [self viewImage];
                             });
                         }
                         else if ([[actionSheet buttonTitleAtIndex:buttonIndex] isEqualToString:@"Change image"]) {
                             NSMutableArray *actions = [NSMutableArray arrayWithObjects:@"Take new", @"Choose from Library", nil];
                             
                             if (FBSession.activeSession.state == FBSessionStateOpen && imagePickerMode == IMAGE_PICKER_MODE_AVATAR)
                                 [actions addObject:@"Import from Facebook"];
                             
                             [UIActionSheet showInView:self.view
                                             withTitle:@"Choose source"
                                     cancelButtonTitle:@"Cancel"
                                destructiveButtonTitle:nil
                                     otherButtonTitles:actions
                                              tapBlock:^(UIActionSheet *actionSheet, NSInteger buttonIndex) {
                                                  if ([[actionSheet buttonTitleAtIndex:buttonIndex] isEqualToString:@"Cancel"])
                                                      return;
                                                  
                                                  if (buttonIndex == 0)
                                                      [self doTakeNew];
                                                  else if (buttonIndex == 1)
                                                      [self doChooseFromGallery];
                                                  else if (buttonIndex == 2)
                                                      [self doImportFromFacebook];
                                              }];
                         }
                         else if ([[actionSheet buttonTitleAtIndex:buttonIndex] isEqualToString:@"Remove image"]) {
                             [DialogBuilder buildYesNoDialogWithTitle:APP_NAME message:@"Are you sure you want to remove this image?" yesTitle:@"Yes" noTitle:@"No" yesBlock:^{
                                 [self changeImage:nil];
                             } noBlock:^{}];
                         }
                     }];
}

#pragma mark - Profile actions

- (void)viewImage {
    UIImageView *i;
    
    if (imagePickerMode == IMAGE_PICKER_MODE_AVATAR)
        i = avatarImage;
    else if (imagePickerMode == IMAGE_PICKER_MODE_COVER)
        i = coverImage;
    
    [EXPhotoViewer showImageFrom:i rootViewController:self.navigationController];
}

- (void)changeImage:(UIImage *)image {
    [[LoadingViewManager getInstance] addLoadingToView:self.navigationController.view withMessage:@"Processing"];
    
    if (imagePickerMode == IMAGE_PICKER_MODE_AVATAR) {
        EditAvatarTask *task = [EditAvatarTask new];
        [task editAvatarWithNewImage:image successBlock:^(ResponseObject *response) {
            [[LoadingViewManager getInstance] removeLoadingView];
            
            user = [Settings getInstance].user;
            
            [self loadAvatar];
        } failureBlock:^(ResponseObject *response) {
            [[LoadingViewManager getInstance] removeLoadingView];
            
            if (response.reason == AUTHORIZATION_NEEDED) {
                [Utils logoutUserAndShowLoginController];
                [Utils showMessage:APP_NAME message:@"Your session has timed out. Please login again."];
            }
            else if (response.reason == DATABASE_ERROR || response.reason == NOT_FOUND || response.reason == NETWORK || response.reason == OTHER)
                [Utils showMessage:APP_NAME message:response.message];
        }];
    }
    else if (imagePickerMode == IMAGE_PICKER_MODE_COVER) {
        EditCoverTask *task = [EditCoverTask new];
        [task editCoverWithNewImage:image successBlock:^(ResponseObject *response) {
            [[LoadingViewManager getInstance] removeLoadingView];
            
            user = [Settings getInstance].user;
            
            [self loadCover];
        } failureBlock:^(ResponseObject *response) {
            [[LoadingViewManager getInstance] removeLoadingView];
            
            if (response.reason == AUTHORIZATION_NEEDED) {
                [Utils logoutUserAndShowLoginController];
                [Utils showMessage:APP_NAME message:@"Your session has timed out. Please login again."];
            }
            else if (response.reason == DATABASE_ERROR || response.reason == NOT_FOUND || response.reason == NETWORK || response.reason == OTHER)
                [Utils showMessage:APP_NAME message:response.message];
        }];
    }
}

#pragma mark - Loading

- (void)loadData {
    [self setTextForCell:firstNameLabel text:user.firstname];
    [self setTextForCell:lastNameLabel text:user.lastname];
    [self setTextForCell:emailLabel text:user.email];
    [self setTextForCell:aboutLabel text:user.about];
    [self setTextForCell:websiteLabel text:user.website];
    
    [self loadAvatar];
    [self loadCover];
}

- (void)setTextForCell:(UILabel *)label text:(NSString *)text {
    label.text = text;
}

- (void)loadAvatar {
    __strong typeof(self) weakSelf = self;
    
    if (user.avatarId > 0) {
        NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:[RequestBuilder buildGetAvatar:user.avatarId]]];
        request.cachePolicy = NSURLRequestReturnCacheDataElseLoad;
        
        [avatarImage setImageWithURLRequest:request placeholderImage:avatarImage.image success:^(NSURLRequest *request, NSHTTPURLResponse *response, UIImage *image) {
            [UIView transitionWithView:weakSelf->avatarImage
                              duration:0.5f
                               options:UIViewAnimationOptionTransitionCrossDissolve
                            animations:^{
                                weakSelf->avatarImage.image = image;
                            } completion:nil];
        } failure:^(NSURLRequest *request, NSHTTPURLResponse *response, NSError *error) {
            weakSelf->avatarImage.image = [UIImage imageNamed:@"default_avatar.jpg"];
        }];
    }
    else
        avatarImage.image = [UIImage imageNamed:@"default_avatar.jpg"];
}

- (void)loadCover {
    __strong typeof(self) weakSelf = self;
    
    if (user.coverId > 0) {
        NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:[RequestBuilder buildGetCover:user.coverId]]];
        request.cachePolicy = NSURLRequestReturnCacheDataElseLoad;
        
        [coverImage setImageWithURLRequest:request placeholderImage:coverImage.image success:^(NSURLRequest *request, NSHTTPURLResponse *response, UIImage *image) {
            [UIView transitionWithView:weakSelf->coverImage
                              duration:0.5f
                               options:UIViewAnimationOptionTransitionCrossDissolve
                            animations:^{
                                weakSelf->coverImage.image = image;
                            } completion:nil];
        } failure:^(NSURLRequest *request, NSHTTPURLResponse *response, NSError *error) {
            weakSelf->coverImage.image = weakSelf->defaultBlurredImage;
        }];
    }
    else {
        if (user.avatarId > 0) { // Darken + blur the user's avatar and use it as cover.
            // Download avatar.
            NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:[RequestBuilder buildGetAvatar:user.avatarId]]];
            request.cachePolicy = NSURLRequestReturnCacheDataElseLoad;
            
            [coverImage setImageWithURLRequest:request placeholderImage:coverImage.image success:^(NSURLRequest *request, NSHTTPURLResponse *response, UIImage *image) {
                [weakSelf setDarkBlurredImageAsCover:image];
            } failure:^(NSURLRequest *request, NSHTTPURLResponse *response, NSError *error) {
                weakSelf->coverImage.image = weakSelf->defaultBlurredImage;
            }];
        }
        else
            weakSelf->coverImage.image = defaultBlurredImage;
    }
}

- (void)setDarkBlurredImageAsCover:(UIImage *)i {
    UIImage *darken = [i colorizeImagWithColor:[UIColor colorWithWhite:0 alpha:0.5]];
    UIImage *blur = [darken blurredImageWithRadius:40 iterations:2 tintColor:nil];
    
    [UIView transitionWithView:coverImage
                      duration:0.5f
                       options:UIViewAnimationOptionTransitionCrossDissolve
                    animations:^{
                        coverImage.image = blur;
                    } completion:nil];
}

#pragma mark - Image picking

- (void)doTakeNew {
    if ( ![UIImagePickerController isSourceTypeAvailable:UIImagePickerControllerSourceTypeCamera] )
        [Utils showMessage:APP_NAME message:@"No camera app was found on this device."];
    else {
        UIImagePickerController *picker = [[UIImagePickerController alloc] init];
        picker.delegate = self;
        picker.allowsEditing = NO;
        picker.sourceType = UIImagePickerControllerSourceTypeCamera;
        picker.mediaTypes = [NSArray arrayWithObjects:
                             (NSString *) kUTTypeImage,
                             nil];
        
        [self presentViewController:picker animated:YES completion:nil];
        picker = nil;
    }
}

- (void)doChooseFromGallery {
    if ( ![UIImagePickerController isSourceTypeAvailable:UIImagePickerControllerSourceTypePhotoLibrary] )
        [Utils showMessage:APP_NAME message:@"No gallery app was found on this device."];
    else {
        galleryPicker = [[UIImagePickerController alloc] init];
        galleryPicker.delegate = self;
        galleryPicker.allowsEditing = NO;
        galleryPicker.sourceType = UIImagePickerControllerSourceTypePhotoLibrary;
        galleryPicker.mediaTypes = [NSArray arrayWithObjects:
                                    (NSString *) kUTTypeImage,
                                    nil];
        
        [self presentViewController:galleryPicker animated:YES completion:nil];
    }
}

- (void)doImportFromFacebook {
    [[LoadingViewManager getInstance] addLoadingToView:self.navigationController.view withMessage:@"Processing"];
    
    [[FBRequest requestForMe] startWithCompletionHandler:
     ^(FBRequestConnection *connection, NSDictionary<FBGraphUser> *aUser, NSError *error) {
         if (!error) {
             NSString *userImageURL = [NSString stringWithFormat:APP_FACEBOOK_AVATAR_URL, [aUser objectID]];
             
             dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
                 // 1.2.1.2.1
                 NSData *imageData = [NSData dataWithContentsOfURL:[NSURL URLWithString:userImageURL]];
                 UIImage *image = [UIImage imageWithData:imageData];
                 
                 dispatch_async(dispatch_get_main_queue(), ^{
                     [[LoadingViewManager getInstance] removeLoadingView];
                     
                     [self changeImage:image];
                 });
             });
         }
         else
             [[LoadingViewManager getInstance] removeLoadingView];
     }];
}

#pragma mark - Navigation

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"SEGUE_EDIT_TEXT_FIELD"]) {
        NSDictionary *senderDict = sender;
        
        EditTextFieldViewController *vc = segue.destinationViewController;
        vc.finishedEditingBlock = senderDict[@"sender"];
        vc.defaultValue = senderDict[@"default"];
        vc.canBeEmpty = [senderDict[@"canBeEmpty"] boolValue];
    }
    else if ([segue.identifier isEqualToString:@"SEGUE_EDIT_TEXT_VIEW"]) {
        NSDictionary *senderDict = sender;
        
        EditTextViewViewController *vc = segue.destinationViewController;
        vc.finishedEditingBlock = senderDict[@"sender"];
        vc.defaultValue = senderDict[@"default"];
        vc.canBeEmpty = [senderDict[@"canBeEmpty"] boolValue];
    }
}

#pragma mark - UIImagePickerControllerDelegate

- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary *)info {
    [picker dismissViewControllerAnimated:YES completion:^{
        UIImage *i = [info objectForKey:UIImagePickerControllerOriginalImage];
        
        UIStoryboard *mainStoryboard = [SlideNavigationController sharedInstance].storyboard;
        UINavigationController *cropperNavigation = [mainStoryboard instantiateViewControllerWithIdentifier:@"ImageCropViewController"];
        ImageCropViewController *cropper = cropperNavigation.viewControllers[0];
        
        cropper.checkBounds = YES;
        cropper.rotateEnabled = YES;
        cropper.doneCallback = ^(UIImage *editedImage, BOOL canceled) {
            [cropperNavigation dismissViewControllerAnimated:YES completion:^{
                if (!canceled)
                    [self changeImage:editedImage];
            }];
        };
        
        cropper.sourceImage = i;
        cropper.previewImage = i;
        
        [self presentViewController:cropperNavigation animated:YES completion:NULL];
        
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0.3 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
            int size = 300;
            cropper.cropRect = CGRectMake(self.view.frame.size.width / 2 - size / 2, (self.view.frame.size.height - (STATUSBAR_HEIGHT + NAVIGATIONBAR_HEIGHT)) / 2 - size / 2, size, size);
            [cropper reset:YES];
        });
    }];
}

- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker {
    [picker dismissViewControllerAnimated:YES completion:NULL];
    picker = nil;
}

#pragma mark - UINavigationControllerDelegate

- (void)navigationController:(UINavigationController *)navigationController willShowViewController:(UIViewController *)viewController animated:(BOOL)animated {
    [[UIApplication sharedApplication] setStatusBarStyle:UIStatusBarStyleLightContent];
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    switch (indexPath.section) {
        case 0: { // Profile+Cover images.
            switch (indexPath.row) {
                case 0: { // Profile image.
                    imagePickerMode = IMAGE_PICKER_MODE_AVATAR;
                    
                    [self onClickImage];
                    
                    break;
                }
                case 1: { // Cover image.
                    imagePickerMode = IMAGE_PICKER_MODE_COVER;
                    
                    [self onClickImage];
                    
                    break;
                }
            }
            
            break;
        }
        case 1: {
            if (indexPath.row == 0) // Edit password
                [self performSegueWithIdentifier:@"SEGUE_EDIT_PASSWORD" sender:nil];
            
            break;
        }
        case 2: { // User info.
            NSString *segue;
            id sender;
            id defaultValue;
            BOOL canBeEmpty = NO;
            
            switch (indexPath.row) {
                case 0: { // First name
                    segue = @"SEGUE_EDIT_TEXT_FIELD";
                    defaultValue = firstNameLabel.text;
                    canBeEmpty = NO;
                    void (^finishBlock)(NSString *) = ^(NSString *text) {
                        firstNameLabel.text = text;
                    };
                    
                    sender = finishBlock;
                    
                    break;
                }
                case 1: { // Last name
                    segue = @"SEGUE_EDIT_TEXT_FIELD";
                    defaultValue = lastNameLabel.text;
                    canBeEmpty = NO;
                    void (^finishBlock)(NSString *) = ^(NSString *text) {
                        lastNameLabel.text = text;
                    };
                    
                    sender = finishBlock;
                    
                    break;
                }
                case 2: { // Email
                    segue = @"SEGUE_EDIT_TEXT_FIELD";
                    defaultValue = emailLabel.text;
                    canBeEmpty = NO;
                    void (^finishBlock)(NSString *) = ^(NSString *text) {
                        emailLabel.text = text;
                    };
                    
                    sender = finishBlock;
                    
                    break;
                }
                case 3: { // About
                    segue = @"SEGUE_EDIT_TEXT_VIEW";
                    defaultValue = aboutLabel.text;
                    canBeEmpty = YES;
                    void (^finishBlock)(NSString *) = ^(NSString *text) {
                        aboutLabel.text = text;
                    };
                    
                    sender = finishBlock;
                    
                    break;
                }
                case 4: { // Website
                    segue = @"SEGUE_EDIT_TEXT_FIELD";
                    defaultValue = websiteLabel.text;
                    canBeEmpty = YES;
                    void (^finishBlock)(NSString *) = ^(NSString *text) {
                        websiteLabel.text = text;
                    };
                    
                    sender = finishBlock;
                    
                    break;
                }
            }
            
            if (segue && sender)
                [self performSegueWithIdentifier:segue sender:@{@"sender":sender, @"default":defaultValue ? defaultValue : @"", @"canBeEmpty":@(canBeEmpty)}];
            
            break;
        }
    }
}

#pragma mark - Setup

- (void)setupTopBar {
    self.title = @"Edit Profile";
    
    UIButton *useButton = [UIButton buttonWithType:UIButtonTypeCustom];
    useButton.frame = CGRectMake(0, 0, 50, 30);
    useButton.layer.cornerRadius = 4;
    [useButton setTitle:@"Save" forState:UIControlStateNormal];
    useButton.backgroundColor = UIColorFromRGB(COLOR_ORANGE);
    [useButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    useButton.titleLabel.font = [UIFont boldSystemFontOfSize:15];
    [useButton addTarget:self action:@selector(onClickSave:) forControlEvents:UIControlEventTouchUpInside];
    
    UIBarButtonItem *negativeSpacer = [[UIBarButtonItem alloc]
                                       initWithBarButtonSystemItem:UIBarButtonSystemItemFixedSpace
                                       target:nil action:nil];
    negativeSpacer.width = -10;
    
    UIBarButtonItem *useItem = [[UIBarButtonItem alloc] initWithCustomView:useButton];
    [self.navigationItem setRightBarButtonItems:@[negativeSpacer, useItem]];
}

- (void)setupDefaultImage {
    UIImage *i = [UIImage imageNamed:@"header_upsidedown.png"];
    UIImage *darken = [i colorizeImagWithColor:[UIColor colorWithWhite:0 alpha:0.5]];
    UIImage *blur = [darken blurredImageWithRadius:40 iterations:2 tintColor:nil];
    
    defaultBlurredImage = blur;
}

- (void)setupImageViews {
    avatarImage.layer.cornerRadius = avatarImage.frame.size.width / 2;
    
    coverImage.image = defaultBlurredImage;
    coverImage.layer.cornerRadius = coverImage.frame.size.width / 2;
}

@end

File: GraffiTab/Classes/ViewControllers/Main/Main/Me/Edit/EditProfileViewController.swift
//
//  EditProfileViewController.swift
//  GraffiTab
//
//  Created by Georgi Christov on 13/04/2016.
//  Copyright © 2016 GraffiTab. All rights reserved.
//

import UIKit
import Alamofire
import GraffiTab_iOS_SDK

enum ImageType {
    case Avatar
    case Cover
}

class EditProfileViewController: BackButtonTableViewController, UIImagePickerControllerDelegate, UINavigationControllerDelegate {

    @IBOutlet weak var avatar: UIImageView!
    @IBOutlet weak var cover: UIImageView!
    @IBOutlet weak var firstnameField: UILabel!
    @IBOutlet weak var lastnameField: UILabel!
    @IBOutlet weak var emailField: UILabel!
    @IBOutlet weak var aboutField: UILabel!
    @IBOutlet weak var websiteField: UILabel!
    
    var user: GTUser = GTSettings.sharedInstance.user!
    var imageType: ImageType?
    let imagePicker = UIImagePickerController()
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        
        setupImageViews()
        setupImagePicker()
        
        loadData()
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        if self.navigationController!.navigationBarHidden {
            self.navigationController?.setNavigationBarHidden(false, animated: true)
        }
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    func onClickSave() {
        let fn = firstnameField.text
        let ln = lastnameField.text
        let e = emailField.text
        var a = aboutField.text
        var w = websiteField.text
        
        if a?.characters.count <= 0 {
            a = nil
        }
        if w?.characters.count <= 0 {
            w = nil
        }
        
        self.view.showActivityViewWithLabel("Processing")
        self.view.rn_activityView.dimBackground = false
        
        GTMeManager.editProfile(fn!, lastName: ln!, email: e!, about: a, website: w, successBlock: { (response) in
            self.view.hideActivityView()
            
            self.user = GTSettings.sharedInstance.user!
            self.loadData()
            
            let delayTime = dispatch_time(DISPATCH_TIME_NOW, Int64(0.3 * Double(NSEC_PER_SEC)))
            dispatch_after(delayTime, dispatch_get_main_queue()) {
                DialogBuilder.showSuccessAlert("Your profile has been changed!", title: App.Title)
            }
        }, failureBlock: { (response) in
            self.view.hideActivityView()
            
            if response.reason == .BadRequest {
                DialogBuilder.showErrorAlert("This email is already used by another user.", title: App.Title)
                return
            }
            
            DialogBuilder.showErrorAlert(response.message, title: App.Title)
        })
    }
    
    // MARK: - Images
    
    func askForImage() {
        UIActionSheet.showInView(self.view, withTitle: "What would you like to do?", cancelButtonTitle: "Cancel", destructiveButtonTitle: "Remove image", otherButtonTitles: ["Choose from gallery", "Take new"], tapBlock: { (actionSheet, index) in
            if index == 0 {
                self.saveImage(nil)
            }
            else if index == 1 {
                self.chooseFromGallery()
            }
            else if index == 2 {
                self.takeNew()
            }
        })
    }
    
    func chooseFromGallery() {
        if UIImagePickerController.isSourceTypeAvailable(.PhotoLibrary) {
            imagePicker.sourceType = .PhotoLibrary
            presentViewController(imagePicker, animated: true, completion: nil)
        }
        else {
            print("DEBUG: Source type .PhotoLibrary not available")
        }
    }
    
    func takeNew() {
        if UIImagePickerController.isSourceTypeAvailable(.Camera) {
            imagePicker.sourceType = .Camera
            presentViewController(imagePicker, animated: true, completion: nil)
        }
        else {
            print("DEBUG: Source type .Camera not available")
        }
    }
    
    func saveImage(image: UIImage?) {
        let avatarSuccessBlock = {
            self.view.hideActivityView()
            
            self.loadAvatar()
            
            let delayTime = dispatch_time(DISPATCH_TIME_NOW, Int64(0.3 * Double(NSEC_PER_SEC)))
            dispatch_after(delayTime, dispatch_get_main_queue()) {
                DialogBuilder.showSuccessAlert("Your avatar has been changed!", title: App.Title)
            }
        }
        let coverSuccessBlock = {
            self.view.hideActivityView()
            
            self.loadCover()
            
            let delayTime = dispatch_time(DISPATCH_TIME_NOW, Int64(0.3 * Double(NSEC_PER_SEC)))
            dispatch_after(delayTime, dispatch_get_main_queue()) {
                DialogBuilder.showSuccessAlert("Your cover has been changed!", title: App.Title)
            }
        }
        
        if image != nil { // Saving a new image.
            if imageType == .Avatar {
                self.view.showActivityViewWithLabel("Processing")
                self.view.rn_activityView.dimBackground = false
                
                GTMeManager.editAvatar(image!, successBlock: { (response) in
                    avatarSuccessBlock()
                }, failureBlock: { (response) in
                    self.view.hideActivityView()
                    
                    DialogBuilder.showErrorAlert(response.message, title: App.Title)
                })
            }
            else {
                self.view.showActivityViewWithLabel("Processing")
                self.view.rn_activityView.dimBackground = false
                
                GTMeManager.editCover(image!, successBlock: { (response) in
                    coverSuccessBlock()
                }, failureBlock: { (response) in
                    self.view.hideActivityView()
                    
                    DialogBuilder.showErrorAlert(response.message, title: App.Title)
                })
            }
        }
        else { // Removing an image.
            if imageType == .Avatar {
                self.view.showActivityViewWithLabel("Processing")
                self.view.rn_activityView.dimBackground = false
                
                GTMeManager.deleteAvatar({ (response) in
                    avatarSuccessBlock()
                }, failureBlock: { (response) in
                    self.view.hideActivityView()
                    
                    DialogBuilder.showErrorAlert(response.message, title: App.Title)
                })
            }
            else {
                self.view.showActivityViewWithLabel("Processing")
                self.view.rn_activityView.dimBackground = false
                
                GTMeManager.deleteCover({ (response) in
                    coverSuccessBlock()
                }, failureBlock: { (response) in
                    self.view.hideActivityView()
                    
                    DialogBuilder.showErrorAlert(response.message, title: App.Title)
                })
            }
        }
    }
    
    // MARK: - Loading
    
    func loadData() {
        firstnameField.text = user.firstName
        lastnameField.text = user.lastName
        emailField.text = user.email
        aboutField.text = user.about
        websiteField.text = user.website
        
        loadAvatar()
        loadCover()
    }
    
    func loadAvatar() {
        if user.avatar != nil {
            Alamofire.request(.GET, user.avatar!.link!)
                .responseImage { response in
                    let image = response.result.value
                    
                    self.avatar.image = image
            }
        }
        else {
            avatar.image = UIImage(named: "default_avatar")!
        }
    }
    
    func loadCover() {
        if user.cover != nil {
            Alamofire.request(.GET, user.cover!.link!)
                .responseImage { response in
                    let image = response.result.value
                    
                    self.cover.image = image
            }
        }
        else {
            cover.image = nil
        }
    }
    
    // MARK: - Navigation

    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if segue.identifier == "SEGUE_EDIT_FIRSTNAME" {
            let vc = segue.destinationViewController as! EditTextFieldViewController
            vc.capitalizationType = .Words
            vc.allowEmpty = false
            vc.defaultValue = firstnameField.text
            vc.doneBlock = { (value) in
                self.firstnameField.text = value
            }
        }
        else if segue.identifier == "SEGUE_EDIT_LASTNAME" {
            let vc = segue.destinationViewController as! EditTextFieldViewController
            vc.capitalizationType = .Words
            vc.allowEmpty = false
            vc.defaultValue = lastnameField.text
            vc.doneBlock = { (value) in
                self.lastnameField.text = value
            }
        }
        else if segue.identifier == "SEGUE_EDIT_EMAIL" {
            let vc = segue.destinationViewController as! EditTextFieldViewController
            vc.keyboardType = .EmailAddress
            vc.allowEmpty = false
            vc.defaultValue = emailField.text
            vc.doneBlock = { (value) in
                self.emailField.text = value
            }
        }
        else if segue.identifier == "SEGUE_EDIT_WEBSITE" {
            let vc = segue.destinationViewController as! EditTextFieldViewController
            vc.keyboardType = .EmailAddress
            vc.allowEmpty = true
            vc.defaultValue = websiteField.text
            vc.doneBlock = { (value) in
                self.websiteField.text = value
            }
        }
        else if segue.identifier == "SEGUE_EDIT_ABOUT" {
            let vc = segue.destinationViewController as! EditTextViewController
            vc.capitalizationType = .Sentences
            vc.allowEmpty = true
            vc.defaultValue = aboutField.text
            vc.doneBlock = { (value) in
                self.aboutField.text = value
            }
        }
    }
    
    // MARK: - UIImagePickerControllerDelegate
    
    func imagePickerController(picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [String : AnyObject]) {
        if let pickedImage = info[UIImagePickerControllerOriginalImage] as? UIImage {
            saveImage(pickedImage)
        }
        
        dismissViewControllerAnimated(true, completion: nil)
    }
    
    // MARK: - UITableViewDelegate
    
    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        tableView.deselectRowAtIndexPath(indexPath, animated: true)
        
        if indexPath.section == 0 {
            if indexPath.row == 0 {
                imageType = .Avatar
                askForImage()
            }
            else {
                imageType = .Cover
                askForImage()
            }
        }
    }
    
    // MARK: - Setup
    
    override func setupTopBar() {
        super.setupTopBar()
        
        self.title = "Edit profile"
        
        let button = UIButton(type: .Custom)
        button.frame = CGRectMake(0, 0, 50, 30)
        button.layer.cornerRadius = 3
        button.setTitle("Save", forState: .Normal)
        button.backgroundColor = UIColor(hexString: Colors.Orange)
        button.setTitleColor(UIColor.whiteColor(), forState: .Normal)
        button.titleLabel?.font = UIFont.boldSystemFontOfSize(15)
        button.addTarget(self, action: #selector(EditProfileViewController.onClickSave), forControlEvents: .TouchUpInside)
        
        let negativeSpacer = UIBarButtonItem(barButtonSystemItem: .FixedSpace, target: nil, action: nil)
        negativeSpacer.width = -10
        
        self.navigationItem.rightBarButtonItems = [negativeSpacer, UIBarButtonItem(customView: button)]
    }
    
    func setupImageViews() {
        avatar.layer.cornerRadius = 5
        cover.layer.cornerRadius = 5
    }
    
    func setupImagePicker() {
        imagePicker.delegate = self
        imagePicker.allowsEditing = true
    }
}
